app/gui/src/project-view/providers/widgetRegistry.ts

```diff
@@ -20,6 +20,17 @@ export namespace WidgetInput {
     }
   }
 
+  /** Returns widget-input data for the given AST tree or token with a specific port ID. */
+  export function FromAstWithPortId<A extends Ast.Ast | Ast.Token>(
+    ast: A,
+    portId: PortId,
+  ): WidgetInput & { value: A } {
+    return {
+      portId,
+      value: ast,
+    }
+  }
```

Comment:
This should be used inside `FromAst` - fewer complex type annotations.

Reply:
Done